### PR TITLE
feat(repo): Pending pegouts controller

### PR DESCRIPTION
**Pull request by rwlockbg**
Related issue: https://github.com/botanix-labs/botanix/issues/757

### CODE REVIEW

**Review by scottmillner — DISMISSED** [review]
thanks for doing this which will prevent that annoying error log now 😃 !

looks like `test_notify_pegout` is failing

File: bin/btc-server/src/server.rs [hunk 1]
```diff
@@ -225,6 +221,16 @@ where
             botanix_height: req.height,
         };
 
+        // do not accept any pending pegouts when there are not available UTXOs
+        let available_utxos =
+            get_available_utxos(&self.pegout_scheduler, &self.db).await.map_err(|e| {
```
Comment (arminsabouri) [inline bin/btc-server/src/server.rs]: its possible that we may pre-maturely reject a pegout even if there are tracked tx i.e change coming into the spendable utxo set. I believe the check should be if there are no available utxos and no tracked txs then there is no possible way there can be pending tx.

**Review by arminsabouri — CHANGES_REQUESTED** [review]

File: bin/btc-server/src/server.rs (hunk 1 above)
Comment (scottmillner) [reply bin/btc-server/src/server.rs]: yeah that makes sense bc the `tracked_tx` could get deeply confirmed and the change returned to the utxo set before the epoch ends and the pegouts are honored

Comment (rwlockbg) [reply bin/btc-server/src/server.rs]: yep, totally! I made the change now.

**Review by scottmillner — DISMISSED** [review]

File: bin/btc-server/src/server.rs [hunk 2]
```diff
@@ -225,6 +221,16 @@ where
...
         };
 
+        // do not accept any pending pegouts when there are not available UTXOs and tracked inputs
+        let (available_utxos, tracked_inputs) =
+            get_available_utxos(&self.pegout_scheduler, &self.db).await.map_err(|e| {
+                error!("Failed to get all available utxos. error: {}", e);
+                internal!("Failed to get all available utxos. error: {}", e)
+            })?;
+        if available_utxos.is_empty() && tracked_inputs.is_empty() {
+            return Ok(tonic::Response::new(rpc::Empty {}));
```
Comment (arminsabouri) [inline bin/btc-server/src/server.rs:232]: Sorry i should have picked up on this earlier. can we log an error here.
"Recieved a pegout request when there are no utxos or pending transactions"

File: bin/btc-server/src/util.rs [hunk 3]
```diff
@@ -408,6 +411,24 @@ pub(crate) fn validate_outputs(psbt: &Psbt, db: &database::Db) -> Result<(), Val
...
+    db: &database::Db,
+) -> Result<(HashMap<OutPoint, Utxo>, HashSet<OutPoint>), database::Error> {
+    let utxos: HashMap<OutPoint, Utxo> =
+        db.iter_utxos().try_fold(HashMap::new(), |mut map, r| {
+            let utxo = r?;
+            map.insert(utxo.outpoint, utxo);
+            Ok::<HashMap<bitcoin::OutPoint, Utxo>, database::Error>(map)
+        })?;
+    // Filter the ones that are still pending and conflict with pending txs.
+    let tracked_inputs = pegout_scheduler.lock().await.tracked_inputs();
```
Comment (arminsabouri) [inline bin/btc-server/src/util.rs]: I think you can just pull the tracked inputs from the db that way you dont need to lock

File: bin/btc-server/src/server.rs (hunk 2 above)
Comment (rwlockbg) [reply bin/btc-server/src/server.rs:232]: done

File: bin/btc-server/src/util.rs (hunk 3 above)
Comment (rwlockbg) [reply bin/btc-server/src/util.rs]: done

**Review by arminsabouri — APPROVED** [review]

**Review by scottmillner — DISMISSED** [review]

**Review by arminsabouri — CHANGES_REQUESTED** [review]
Looks like some unit tests in btc server are failing for legit reasons

**Comment by rwlockbg** [comment]
> Looks like some unit tests in btc server are failing for legit reasons

Yes, the notify pegouts is failing for legit reasons. Maybe we should think about whether this code application is always correct. In the case of pegouts there are 0 tracked txs in the db hence its failing. Suggestions ?

**Comment by arminsabouri** [comment]
> > Looks like some unit tests in btc server are failing for legit reasons
> 
> Yes, the notify pegouts is failing for legit reasons. Maybe we should think about whether this code application is always correct. In the case of pegouts there are 0 tracked txs in the db hence its failing. Suggestions ?

@rwlockbg to get around this in the test, I would just add some random utxos to the db before we notify the pegout. And I would also add a test condition to the `test_notify_pegouts` that ensures no pending pegouts are added when there are no utxos available.

**Review by arminsabouri — APPROVED** [review]

**Pull request merged**